akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala

```diff
@@ -50,115 +49,120 @@ import akka.util.ByteString
-            else Some(trimmed)
-          }
-      }
-  }
+  def poll(): Option[ByteString] =
+    try {
+      val foundObject = seekObject()
+      if (!foundObject) None
+      else
+        (pos: @switch) match {
+          case -1 | 0 => None
+          case _ =>
+            if (start == pos) None
+            else {
+              val res = ByteString.fromArrayUnsafe(buffer, start, pos - start)
+              start = pos
+              Some(res)
+            }
+        }
+    } catch {
+      case _: ArithmeticException =>
+        throw new FramingException(s"Invalid JSON encountered at position [$pos] of [${ByteString(buffer).utf8String}]")
+    }
 
   /** @return true if an entire valid JSON object was found, false otherwise */
   private def seekObject(): Boolean = {
     completedObject = false
-    val bufSize = buffer.size
-    while (pos != -1 && (pos < bufSize && pos < maximumObjectLength) && !completedObject) proceed(buffer(pos))
+    val bufSize = buffer.length
+
+    skipToNextObject(bufSize)
+
+    while (pos < bufSize && pos < maximumObjectLength && !completedObject) {
+      val input = buffer(pos)
+      proceed(input)
+      pos += 1
+    }
 
     if (pos >= maximumObjectLength)
       throw new FramingException(s"""JSON element exceeded maximumObjectLength ($maximumObjectLength bytes)!""")
 
     completedObject
   }
 
-  private def proceed(input: Byte): Unit = {
-    if (input == SquareBraceStart && outsideObject) {
-      // outer object is an array
-      pos += 1
-      trimFront += 1
-    } else if (input == SquareBraceEnd && outsideObject) {
-      // outer array completed!
-      pos += 1
-      trimFront += 1
-    } else if (input == Comma && outsideObject) {
-      // do nothing
-      pos += 1
-      trimFront += 1
-    } else if (input == Backslash) {
-      if (lastInput == Backslash & isStartOfEscapeSequence) isStartOfEscapeSequence = false
-      else isStartOfEscapeSequence = true
-      pos += 1
-    } else if (input == DoubleQuote) {
-      if (!isStartOfEscapeSequence) inStringExpression = !inStringExpression
-      isStartOfEscapeSequence = false
-      pos += 1
-    } else if (input == CurlyBraceStart && !inStringExpression) {
-      isStartOfEscapeSequence = false
-      depth += 1
-      pos += 1
-    } else if (input == CurlyBraceEnd && !inStringExpression) {
-      isStartOfEscapeSequence = false
-      depth -= 1
-      pos += 1
-      if (depth == 0) completedObject = true
-    } else if (isWhitespace(input) && !inStringExpression) {
-      pos += 1
-      if (depth == 0) trimFront += 1
-    } else if (insideObject) {
-      isStartOfEscapeSequence = false
+  private def skipToNextObject(bufSize: Int): Unit =
+    while (pos != -1 && pos < bufSize && pos < maximumObjectLength && depth == 0) {
+      val outer = outerChars(buffer(pos) & 0xff)
+      start += outer & 1
+      depth = (outer & 2) >> 1
+
+      1 / outer // HACK: use arithmetic exception to cover error case
```

Comment:
Could we put the try-catch-rethrow here instead, closer to the hack, so to speak?

Reply:
It's quite sensible to where we put this, it might work putting it on the `seekObject` method, but better to avoid it near the inner loops.